apiserver/facades/agent/credentialvalidator/credentialvalidator.go

```diff
@@ -0,0 +1,142 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+package credentialvalidator
+
+import (
+	"github.com/juju/errors"
+	"github.com/juju/loggo"
+	"gopkg.in/juju/names.v2"
+
+	"github.com/juju/juju/apiserver/common"
+	"github.com/juju/juju/apiserver/facade"
+	"github.com/juju/juju/apiserver/params"
+	"github.com/juju/juju/state"
+	"github.com/juju/juju/state/watcher"
+)
+
+var logger = loggo.GetLogger("juju.api.credentialvalidator")
+
+// CredentialValidator defines the methods on credentialvalidator API endpoint.
+type CredentialValidator interface {
+	ModelCredentials(params.Entities) (params.ModelCredentialResults, error)
+	WatchCredential(params.Entities) (params.NotifyWatchResults, error)
+}
+
+type CredentialValidatorAPI struct {
+	backend   Backend
+	resources facade.Resources
+}
+
+var _ CredentialValidator = (*CredentialValidatorAPI)(nil)
+
+// NewCredentialValidatorAPI creates a new CredentialValidator API endpoint on server-side.
+func NewCredentialValidatorAPI(st *state.State, resources facade.Resources, authorizer facade.Authorizer) (*CredentialValidatorAPI, error) {
+	return internalNewCredentialValidatorAPI(NewStateBackend(st), resources, authorizer)
+}
+
+func internalNewCredentialValidatorAPI(backend Backend, resources facade.Resources, authorizer facade.Authorizer) (*CredentialValidatorAPI, error) {
+	hostAuthTag := authorizer.GetAuthTag()
+	if hostAuthTag == nil {
+		return nil, common.ErrPerm
+	}
+
+	return &CredentialValidatorAPI{
+		resources: resources,
+		backend:   backend,
+	}, nil
+}
+
+// auth only accepts the model tag reported by the backend.
+func (api *CredentialValidatorAPI) auth(tagString string) error {
+	tag, err := names.ParseModelTag(tagString)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if tag.Id() != api.backend.ModelUUID() {
+		return common.ErrPerm
+	}
+	return nil
+}
+
+// WatchCredential returns a collection of NotifyWatchers that observe
+// changes to the given cloud credentials.
+// The order of returned watchers is important and corresponds directly to the
+// order of supplied cloud credentials collection.
+func (api *CredentialValidatorAPI) WatchCredential(args params.Entities) (params.NotifyWatchResults, error) {
```

Comment:
This method should be called "WatchCredentials" as it takes args for > 1 credential